File: ios/ShiLv/Services/TripDetector.swift
import Foundation
@preconcurrency import CoreLocation

struct TripDetectionConfiguration: Sendable {
    var minimumDistanceFromHome: CLLocationDistance = 80_000
    var maximumGapWithinTrip: TimeInterval = 36 * 60 * 60
    var minimumLocatedPhotos: Int = 5
    var minimumTotalPhotos: Int = 12
    var eventTimeGap: TimeInterval = 2 * 60 * 60
    var eventDistanceGap: CLLocationDistance = 3_000
}

struct TripDetector: Sendable {
    let configuration: TripDetectionConfiguration
    private let calendar = Calendar.current

    init(configuration: TripDetectionConfiguration = .init()) {
        self.configuration = configuration
    }

    func detect(from photos: [PhotoRecord], now: Date = Date()) -> ScanSnapshot {
        if Task.isCancelled { return ScanSnapshot(version: 1, scannedAt: now, accessiblePhotoCount: photos.count, locatedPhotoCount: 0, inferredHome: nil, trips: []) }
        let sorted = photos.sorted { $0.creationDate < $1.creationDate }
        let eligible = sorted.filter { $0.isScreenshot != true }
        let located = eligible.filter { $0.location != nil }
        let home = inferHome(from: located)
        guard let home else {
            return ScanSnapshot(version: 1, scannedAt: now, accessiblePhotoCount: sorted.count, locatedPhotoCount: located.count, inferredHome: nil, trips: [])
        }

        let locatedGroups = splitAwayAnchors(located, home: home)
        var trips: [DiscoveredTrip] = []

        for group in locatedGroups where group.count >= configuration.minimumLocatedPhotos {
            if Task.isCancelled { break }
            guard let first = group.first, let last = group.last else { continue }
            let paddedStart = first.creationDate.addingTimeInterval(-6 * 60 * 60)
            let paddedEnd = last.creationDate.addingTimeInterval(6 * 60 * 60)
            let allTripPhotos = eligible.filter { $0.creationDate >= paddedStart && $0.creationDate <= paddedEnd }
            guard allTripPhotos.count >= configuration.minimumTotalPhotos else { continue }
            let uniqueDays = Set(allTripPhotos.map { calendar.startOfDay(for: $0.creationDate) })
            guard uniqueDays.count >= 2 || group.count >= 30 else { continue }

            let points = group.compactMap(\.location)
            let center = points.centroid
            let distance = center?.distance(to: home) ?? configuration.minimumDistanceFromHome
            let days = buildDays(from: allTripPhotos)
            trips.append(DiscoveredTrip(
                id: stableTripID(start: first.creationDate, end: last.creationDate),
                title: defaultTripTitle(start: first.creationDate, end: last.creationDate),
                startDate: first.creationDate,
                endDate: last.creationDate,
                photoIDs: allTripPhotos.map(\.id),
                days: days,
                center: center,
                distanceFromHomeMeters: distance,
                isConfirmed: false,
                isHidden: false,
                detectedAt: now,
                summary: nil,
                isFavorite: false,
                coverPhotoIDOverride: nil
            ))
        }

        return ScanSnapshot(
            version: 1,
            scannedAt: now,
            accessiblePhotoCount: sorted.count,
            locatedPhotoCount: located.count,
            inferredHome: home,
            trips: mergeOverlapping(trips).sorted { $0.startDate > $1.startDate }
        )
    }

    func inferHome(from photos: [PhotoRecord]) -> GeoPoint? {
        let located = photos.compactMap { photo -> (GeoPoint, Date)? in
            guard let location = photo.location else { return nil }
            return (location, photo.creationDate)
        }
        guard located.count >= configuration.minimumLocatedPhotos else { return nil }

        struct Cell: Hashable { let lat: Int; let lon: Int }
        var buckets: [Cell: [(GeoPoint, Date)]] = [:]
        for (point, date) in located {
            let cell = Cell(lat: Int((point.latitude * 5).rounded()), lon: Int((point.longitude * 5).rounded()))
            buckets[cell, default: []].append((point, date))
        }
        let best = buckets.max { lhs, rhs in
            let lhsDays = Set(lhs.value.map { calendar.startOfDay(for: $0.1) }).count
            let rhsDays = Set(rhs.value.map { calendar.startOfDay(for: $0.1) }).count
            return lhsDays == rhsDays ? lhs.value.count < rhs.value.count : lhsDays < rhsDays
        }?.value
        return best?.map(\.0).centroid
    }

    private func splitAwayAnchors(_ locatedPhotos: [PhotoRecord], home: GeoPoint) -> [[PhotoRecord]] {
        var groups: [[PhotoRecord]] = []
        var current: [PhotoRecord] = []
        let sparseTravelGap = max(configuration.maximumGapWithinTrip, 72 * 60 * 60)

        for photo in locatedPhotos {
            guard let point = photo.location else { continue }
            let isAway = point.distance(to: home) >= configuration.minimumDistanceFromHome
            if !isAway {
                if !current.isEmpty { groups.append(current); current = [] }
                continue
            }
            if let previous = current.last, photo.creationDate.timeIntervalSince(previous.creationDate) > sparseTravelGap {
                groups.append(current); current = []
            }
            current.append(photo)
        }
        if !current.isEmpty { groups.append(current) }
        return groups
    }

    private func buildDays(from photos: [PhotoRecord]) -> [TravelDay] {
        let byDay = Dictionary(grouping: photos) { calendar.startOfDay(for: $0.creationDate) }
        return byDay.keys.sorted().enumerated().compactMap { index, date in
            guard let photosForDay = byDay[date] else { return nil }
            let events = buildEvents(from: photosForDay.sorted { $0.creationDate < $1.creationDate })
            return TravelDay(id: stableDayID(date), date: date, title: "第 \(index + 1) 天", events: events)
        }
    }

    private func buildEvents(from photos: [PhotoRecord]) -> [MemoryEvent] {
        guard let first = photos.first else { return [] }
        var groups = [[first]]
        for photo in photos.dropFirst() {
            guard let previous = groups.last?.last else { continue }
            let timeGap = photo.creationDate.timeIntervalSince(previous.creationDate)
            let distance = zip(previous.location, photo.location).map { pair in pair.0.distance(to: pair.1) } ?? 0
            if timeGap > configuration.eventTimeGap || distance > configuration.eventDistanceGap {
                groups.append([photo])
            } else {
                groups[groups.count - 1].append(photo)
            }
        }
        return groups.compactMap { group in
            guard let first = group.first, let last = group.last else { return nil }
            let points = group.compactMap(\.location)
            let start = first.creationDate
            return MemoryEvent(
                id: stableEventID(start, firstPhotoID: first.id),
                title: eventTitle(for: start),
                startDate: start,
                endDate: last.creationDate,
                photoIDs: group.map(\.id),
                location: points.centroid,
                placeName: nil,
                note: "",
                isHidden: false,
                summary: nil,
                featuredPhotoIDs: representativePhotoIDs(from: group),
                coverPhotoIDOverride: nil,
                excludedPhotoIDs: nil
            )
        }
    }

    private func representativePhotoIDs(from photos: [PhotoRecord]) -> [String] {
        let candidates = photos.filter { $0.isScreenshot != true }
        let source = candidates.isEmpty ? photos : candidates
        guard source.count > 5 else { return source.map(\.id) }
        var indices = [0, source.count / 4, source.count / 2, source.count * 3 / 4, source.count - 1]
        if let favorite = source.firstIndex(where: \.isFavorite) { indices[2] = favorite }
        return indices.map { source[$0].id }.reduce(into: [String]()) { result, id in
            if !result.contains(id) { result.append(id) }
        }
    }

    private func mergeOverlapping(_ trips: [DiscoveredTrip]) -> [DiscoveredTrip] {
        let sorted = trips.sorted { $0.startDate < $1.startDate }
        guard var current = sorted.first else { return [] }
        var result: [DiscoveredTrip] = []
        for next in sorted.dropFirst() {
            if next.startDate <= current.endDate.addingTimeInterval(12 * 60 * 60) {
                let photoIDs = (current.photoIDs + next.photoIDs).reduce(into: [String]()) { result, id in if !result.contains(id) { result.append(id) } }
                let allDays = mergeDays(current.days + next.days)
                current = DiscoveredTrip(id: current.id, title: current.title, startDate: current.startDate, endDate: max(current.endDate, next.endDate), photoIDs: photoIDs, days: allDays, center: [current.center, next.center].compactMap { $0 }.centroid, distanceFromHomeMeters: max(current.distanceFromHomeMeters, next.distanceFromHomeMeters), isConfirmed: current.isConfirmed, isHidden: false, detectedAt: current.detectedAt, summary: current.summary, isFavorite: current.isFavorite, coverPhotoIDOverride: current.coverPhotoIDOverride, suppressedEventIDs: current.suppressedEventIDs)
            } else {
                result.append(current); current = next
            }
        }
        result.append(current)
        return result
    }

    private func mergeDays(_ days: [TravelDay]) -> [TravelDay] {
        let grouped = Dictionary(grouping: days) { calendar.startOfDay(for: $0.date) }
        return grouped.keys.sorted().map { date in
            let events = (grouped[date] ?? []).flatMap(\.events).sorted { $0.startDate < $1.startDate }
            return TravelDay(id: stableDayID(date), date: date, title: grouped[date]?.first?.title ?? "旅途中的一天", events: events)
        }
    }

    private func defaultTripTitle(start: Date, end: Date) -> String {
        let formatter = DateFormatter(); formatter.locale = Locale(identifier: "zh_CN"); formatter.dateFormat = "M月旅行"
        return formatter.string(from: start)
    }

    private func eventTitle(for date: Date) -> String {
        switch calendar.component(.hour, from: date) {
        case 0..<6: return "夜里的片段"
        case 6..<11: return "上午的记忆"
        case 11..<14: return "午间停留"
        case 14..<18: return "下午的漫步"
        default: return "傍晚与夜晚"
        }
    }

    private func stableTripID(start: Date, end: Date) -> UUID { stableUUID("trip-\(start.timeIntervalSince1970)-\(end.timeIntervalSince1970)") }
    private func stableDayID(_ date: Date) -> UUID { stableUUID("day-\(date.timeIntervalSince1970)") }
    private func stableEventID(_ date: Date, firstPhotoID: String) -> UUID { stableUUID("event-\(date.timeIntervalSince1970)-\(firstPhotoID)") }

    private func stableUUID(_ value: String) -> UUID {
        var bytes = [UInt8](repeating: 0, count: 16)
        for (index, byte) in value.utf8.enumerated() { bytes[index % 16] = bytes[index % 16] &* 31 &+ byte }
        bytes[6] = (bytes[6] & 0x0F) | 0x40; bytes[8] = (bytes[8] & 0x3F) | 0x80
        return UUID(uuid: (bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7], bytes[8], bytes[9], bytes[10], bytes[11], bytes[12], bytes[13], bytes[14], bytes[15]))
    }
}

private func zip<T, U>(_ lhs: T?, _ rhs: U?) -> (T, U)? {
    guard let lhs, let rhs else { return nil }
    return (lhs, rhs)
}

File: android/app/src/main/java/com/example/shilv/domain/TripDetector.kt
package com.example.shilv.domain

import com.example.shilv.data.DiscoveredTrip
import com.example.shilv.data.GeoPoint
import com.example.shilv.data.MemoryEvent
import com.example.shilv.data.PhotoRecord
import com.example.shilv.data.ScanSnapshot
import com.example.shilv.data.TravelDay
import com.example.shilv.data.TripDates
import com.example.shilv.data.centroid
import java.util.UUID

/**
 * 旅行检测器：从照片元数据中还原“一次旅行”。
 * 逻辑与 iOS TripDetector 完全对齐：
 *  - 推断常驻区域（home）
 *  - 将远离 home 且时间连续的带定位照片切分为候选组
 *  - 补入无定位照片、按天/按事件聚类
 */
class TripDetector(
    private val minimumDistanceFromHome: Double = 80_000.0,
    private val maximumGapWithinTrip: Long = 36 * 60 * 60 * 1000L,
    private val minimumLocatedPhotos: Int = 5,
    private val minimumTotalPhotos: Int = 12,
    private val eventTimeGap: Long = 2 * 60 * 60 * 1000L,
    private val eventDistanceGap: Double = 3_000.0,
) {
    fun detect(photos: List<PhotoRecord>, now: Long = System.currentTimeMillis()): ScanSnapshot {
        val sorted = photos.sortedBy { it.creationDate }
        val eligible = sorted.filter { !it.isScreenshot }
        val located = eligible.filter { it.location != null }
        val home = inferHome(located)
        if (home == null) {
            return ScanSnapshot(
                version = 1, scannedAt = now, accessiblePhotoCount = sorted.size,
                locatedPhotoCount = located.size, inferredHome = null, trips = emptyList(),
            )
        }

        val locatedGroups = splitAwayAnchors(located, home)
        val trips = mutableListOf<DiscoveredTrip>()

        for (group in locatedGroups) {
            if (group.size < minimumLocatedPhotos) continue
            val first = group.first()
            val last = group.last()
            val paddedStart = first.creationDate - 6 * 60 * 60 * 1000L
            val paddedEnd = last.creationDate + 6 * 60 * 60 * 1000L
            val allTripPhotos = eligible.filter { it.creationDate in paddedStart..paddedEnd }
            if (allTripPhotos.size < minimumTotalPhotos) continue
            val uniqueDays = allTripPhotos.map { TripDates.startOfDay(it.creationDate) }.toSet()
            if (uniqueDays.size < 2 && group.size < 30) continue

            val points = group.mapNotNull { it.location }
            val center = points.centroid()
            val distance = center?.distance(home) ?: minimumDistanceFromHome
            val days = buildDays(allTripPhotos)
            trips.add(
                DiscoveredTrip(
                    id = stableTripID(first.creationDate, last.creationDate),
                    title = defaultTripTitle(first.creationDate),
                    startDate = first.creationDate,
                    endDate = last.creationDate,
                    photoIDs = allTripPhotos.map { it.id },
                    days = days,
                    center = center,
                    distanceFromHomeMeters = distance,
                    isConfirmed = false,
                    isHidden = false,
                    detectedAt = now,
                ),
            )
        }

        return ScanSnapshot(
            version = 1, scannedAt = now, accessiblePhotoCount = sorted.size,
            locatedPhotoCount = located.size, inferredHome = home,
            trips = mergeOverlapping(trips).sortedByDescending { it.startDate },
        )
    }

    fun inferHome(photos: List<PhotoRecord>): GeoPoint? {
        val located = photos.mapNotNull { p -> p.location?.let { it to p.creationDate } }
        if (located.size < minimumLocatedPhotos) return null
        data class Cell(val lat: Int, val lon: Int)
        val buckets = mutableMapOf<Cell, MutableList<Pair<GeoPoint, Long>>>()
        for ((point, date) in located) {
            val cell = Cell((point.latitude * 5).toInt(), (point.longitude * 5).toInt())
            buckets.getOrPut(cell) { mutableListOf() }.add(point to date)
        }
        val best = buckets.maxWithOrNull { lhs, rhs ->
            val lhsDays = lhs.value.map { TripDates.startOfDay(it.second) }.toSet().size
            val rhsDays = rhs.value.map { TripDates.startOfDay(it.second) }.toSet().size
            if (lhsDays == rhsDays) lhs.value.size.compareTo(rhs.value.size) else lhsDays.compareTo(rhsDays)
        }?.value
        return best?.map { it.first }?.centroid()
    }

    private fun splitAwayAnchors(locatedPhotos: List<PhotoRecord>, home: GeoPoint): List<List<PhotoRecord>> {
        val groups = mutableListOf<List<PhotoRecord>>()
        var current = mutableListOf<PhotoRecord>()
        val sparseTravelGap = maxOf(maximumGapWithinTrip, 72 * 60 * 60 * 1000L)
        for (photo in locatedPhotos) {
            val point = photo.location ?: continue
            val isAway = point.distance(home) >= minimumDistanceFromHome
            if (!isAway) {
                if (current.isNotEmpty()) { groups.add(current); current = mutableListOf() }
                continue
            }
            val previous = current.lastOrNull()
            if (previous != null && photo.creationDate - previous.creationDate > sparseTravelGap) {
                groups.add(current); current = mutableListOf()
            }
            current.add(photo)
        }
        if (current.isNotEmpty()) groups.add(current)
        return groups
    }

    private fun buildDays(photos: List<PhotoRecord>): List<TravelDay> {
        val byDay = photos.groupBy { TripDates.startOfDay(it.creationDate) }
        return byDay.keys.sorted()
            .mapIndexed { index, date ->
                val dayPhotos = byDay[date]!!.sortedBy { it.creationDate }
                TravelDay(
                    id = stableDayID(date),
                    date = date,
                    title = "第 ${index + 1} 天",
                    events = buildEvents(dayPhotos),
                )
            }
    }

    private fun buildEvents(photos: List<PhotoRecord>): List<MemoryEvent> {
        val first = photos.firstOrNull() ?: return emptyList()
        val groups = mutableListOf<MutableList<PhotoRecord>>(mutableListOf(first))
        for (photo in photos.drop(1)) {
            val previous = groups.last().last()
            val timeGap = photo.creationDate - previous.creationDate
            val distance = previous.location?.let { a -> photo.location?.let { a.distance(it) } } ?: 0.0
            if (timeGap > eventTimeGap || distance > eventDistanceGap) {
                groups.add(mutableListOf(photo))
            } else {
                groups.last().add(photo)
            }
        }
        return groups.mapNotNull { group ->
            val gFirst = group.first()
            val gLast = group.last()
            val points = group.mapNotNull { it.location }
            val start = gFirst.creationDate
            MemoryEvent(
                id = stableEventID(start, gFirst.id),
                title = eventTitle(start),
                startDate = start,
                endDate = gLast.creationDate,
                photoIDs = group.map { it.id },
                location = points.centroid(),
                note = "",
                isHidden = false,
                featuredPhotoIDs = representativePhotoIDs(group),
            )
        }
    }

    private fun representativePhotoIDs(photos: List<PhotoRecord>): List<String> {
        val candidates = photos.filter { !it.isScreenshot }
        val source = if (candidates.isEmpty()) photos else candidates
        if (source.size <= 5) return source.map { it.id }
        var indices = listOf(0, source.size / 4, source.size / 2, source.size * 3 / 4, source.size - 1)
        val favoriteIndex = source.indexOfFirst { it.isFavorite }
        if (favoriteIndex >= 0) indices = indices.mapIndexed { i, v -> if (i == 2) favoriteIndex else v }
        return indices.map { source[it].id }.distinct()
    }

    private fun mergeOverlapping(trips: List<DiscoveredTrip>): List<DiscoveredTrip> {
        val sorted = trips.sortedBy { it.startDate }
        val current = sorted.firstOrNull() ?: return emptyList()
        val result = mutableListOf<DiscoveredTrip>()
        var cur = current
        for (next in sorted.drop(1)) {
            if (next.startDate <= cur.endDate + 12 * 60 * 60 * 1000L) {
                val photoIDs = (cur.photoIDs + next.photoIDs).distinct()
                val allDays = mergeDays(cur.days + next.days)
                cur = DiscoveredTrip(
                    id = cur.id, title = cur.title, startDate = cur.startDate,
                    endDate = maxOf(cur.endDate, next.endDate), photoIDs = photoIDs, days = allDays,
                    center = listOfNotNull(cur.center, next.center).centroid(),
                    distanceFromHomeMeters = maxOf(cur.distanceFromHomeMeters, next.distanceFromHomeMeters),
                    isConfirmed = cur.isConfirmed, isHidden = false, detectedAt = cur.detectedAt,
                    summary = cur.summary, isFavorite = cur.isFavorite,
                    coverPhotoIDOverride = cur.coverPhotoIDOverride,
                    suppressedEventIDs = cur.suppressedEventIDs,
                )
            } else {
                result.add(cur); cur = next
            }
        }
        result.add(cur)
        return result
    }

    private fun mergeDays(days: List<TravelDay>): List<TravelDay> {
        val grouped = days.groupBy { TripDates.startOfDay(it.date) }
        return grouped.keys.sorted().map { date ->
            val events = grouped[date].orEmpty().flatMap { it.events }.sortedBy { it.startDate }
            TravelDay(
                id = stableDayID(date),
                date = date,
                title = grouped[date]?.firstOrNull()?.title ?: "旅途中的一天",
                events = events,
            )
        }
    }

    private fun defaultTripTitle(start: Long): String = "${TripDates.monthOf(start)}月旅行"

    private fun eventTitle(date: Long): String = when (TripDates.hourOf(date)) {
        in 0 until 6 -> "夜里的片段"
        in 6 until 11 -> "上午的记忆"
        in 11 until 14 -> "午间停留"
        in 14 until 18 -> "下午的漫步"
        else -> "傍晚与夜晚"
    }

    private fun stableTripID(start: Long, end: Long): String = stableUUID("trip-$start-$end")
    private fun stableDayID(date: Long): String = stableUUID("day-$date")
    private fun stableEventID(date: Long, firstPhotoID: String): String = stableUUID("event-$date-$firstPhotoID")

    private fun stableUUID(value: String): String {
        val bytes = ByteArray(16)
        val utf8 = value.toByteArray(Charsets.UTF_8)
        for ((index, byte) in utf8.withIndex()) bytes[index % 16] = (bytes[index % 16] * 31 + byte).toByte()
        bytes[6] = ((bytes[6].toInt() and 0x0F) or 0x40).toByte()
        bytes[8] = ((bytes[8].toInt() and 0x3F) or 0x80).toByte()
        val msb = java.nio.ByteBuffer.wrap(bytes, 0, 8).long
        val lsb = java.nio.ByteBuffer.wrap(bytes, 8, 8).long
        return UUID(msb, lsb).toString()
    }
}